ex_cubic_ingestion/lib/ex_cubic_ingestion/workers/fetch_dmap.ex

```
@@ -47,14 +47,24 @@ defmodule ExCubicIngestion.Workers.FetchDmap do
   end
 
   @doc """
-  Construct the full URL to the feed, applying some overriding logic for
-  last updated (if passed in).
+  Construct the full URL to the feed, with the correct API key, and applying some
+  overriding logic for last updated (if passed in).
   """
   @spec construct_feed_url(CubicDmapFeed.t(), String.t() | nil) :: String.t()
   def construct_feed_url(feed_rec, last_updated) do
     dmap_base_url = Application.fetch_env!(:ex_cubic_ingestion, :dmap_base_url)
 
-    dmap_api_key = Application.fetch_env!(:ex_cubic_ingestion, :dmap_api_key)
+    dmap_api_key =
+      cond do
+        String.starts_with?(feed_rec.relative_url, "/datasetcontrolleduserapi") ->
+          Application.fetch_env!(:ex_cubic_ingestion, :dmap_controlled_user_api_key)
+
+        String.starts_with?(feed_rec.relative_url, "/datasetpublicusersapi") ->
+          Application.fetch_env!(:ex_cubic_ingestion, :dmap_public_user_api_key)
+
+        true ->
+          ""
```

question: should this crash rather than using an empty API key?

---

Sure. Are you thinking something like `raise "No API key available."`?

---

that's perfect!